### PR TITLE
Add KEEP_UNARY_IN_INDIVIDUALS option to simplify()

### LINKED ISSUE #1113: Keeping specific unary nodes though simplify
There are some cases where, during simplify, we might want to leave specific unary nodes in the TS, as long as they are on the lineage to a sample node. At the moment, we can only retain unary nodes by a. specifying `keep_unary=T` in which case *all* unary nodes are kept but risk being lost on the next simplify or b. marking the node as a "sample node" but then the node is always retained even if it is not on the lineage from another sample node. The specific example I have encountered with @mmosborn is where a "census" is performed at a certain time, and the nodes at that time, *and the individuals associated with them*, need to be retained as long as they are in the ancestry. This allows us to get information about individuals and nodes in the past (for example, the spatial location of the individual) even though the nodes aren't strictly "samples" because we don't think we have full genomic information about them.

@jeromekelleher suggests that a way to do this is via an option to simplify which says "keep unary nodes associated with an individual, unless they aren't on a path from a sample". I find that a little contorted, but my suggestion of flagging up the required unary nodes using a flag like NODE_IS_SAMPLE has other problems because it associates a permanent state with the node in question. Other ideas welcome below.

### PR DESCRIPTION
**Pull request by hyanwong**
## Description

Adds KEEP_UNARY_IN_INDIVIDUALS option to simplify()

Fixes #1113 

# PR Checklist:

- [x] Tests that fully cover new/changed functionality.
- [x] Documentation including tutorial content if appropriate.
- [x] Changelogs, if there are API changes.

### CODE REVIEW

**Comment by hyanwong** [comment]
If we want to get this into a C API release version so we can release SLiM 3.5.1, what do we need to do @benjeffery ?

**Comment by benjeffery** [comment]
> If we want to get this into a C API release version so we can release SLiM 3.5.1, what do we need to do @benjeffery ?

I think we can push out a C API release soon after this is merged - we can hold #866 till after as it is still in progress anyway,

File: c/tskit/tables.h
Comment (jeromekelleher) [inline c/tskit/tables.h:2677]: only if

**Review by jeromekelleher — APPROVED** [review]
LGTM! One nit.

File: c/tskit/tables.h
Comment (hyanwong) [reply c/tskit/tables.h]: Corrected. Well spotted.

**Comment by hyanwong** [comment]
> LGTM! One nit.

Thanks. I didn't know if we wanted to reorder the bits passed as flags to `simplify()` so that the two TSK_KEEP_UNARY bits were together? I don't know if the actual bit used is purely an internal thing or not, but I though better not to rearrange for the moment?

File: c/tskit/tables.c
Comment (petrelharp) [inline c/tskit/tables.c:6097]: "indivifuals"

**Comment by jeromekelleher** [comment]
> I don't know if the actual bit used is purely an internal thing or not, but I though better not to rearrange for the moment?

I don't think there's much point in changing the bit values - someone might be depending on them, and there isn't really much to be gained from having them contiguous (and when we tag 1.0 we'll have to get used to these compromises anyway).

**Comment by petrelharp** [comment]
LGTM, and it's very straightforward, although how do we feel about it not really being tested for correctness? I'd say that this implementation is "obviously correct", but changes down the line to simplify's algorithm might break that.

File: c/tskit/tables.c
Comment (hyanwong) [reply c/tskit/tables.c:6097]: Oops 😊

**Comment by hyanwong** [comment]
> LGTM, and it's very straightforward, although how do we feel about it not really being tested for correctness? I'd say that this implementation is "obviously correct", but changes down the line to simplify's algorithm might break that.

Correctness-wise don't think it's much different from KEEP_UNARY, is it? There are a few basic tests I added in the suite, but I imagine this might get tested more thoroughly (? for correctness) if it is implemented in the Python API too?

**Comment by petrelharp** [comment]
> I imagine this might get tested more thoroughly (? for correctness) if it is implemented in the Python API too?

Right, I'm wondering if we need to actually do this.

**Comment by hyanwong** [comment]
> > I imagine this might get tested more thoroughly (? for correctness) if it is implemented in the Python API too?
> 
> Right, I'm wondering if we need to actually do this.

I think not, because the new functionality is simply a slightly restricted version of keep_unary. I'm not sure how it could go seriously wrong?

**Comment by petrelharp** [comment]
Right now, I agree - I don't see how it could go wrong. But, suppose that next month, someone else makes some other change to simplify that breaks this. They don't catch it becaue they add tests for the functionality that they've implemented, and assume that everything else (e.g., this) is well-tested elsewhere.

**Comment by hyanwong** [comment]
But I have tested a (admittedly single tree) case where one unary node has an individual and another one doesn't, and I'd checked the number of nodes that are expected to remain after the KEEP_UNARY_IN_INDIVIDUALS flag. What other tests might you be thinking of?

**Comment by jeromekelleher** [comment]
I think the ``test_simplest_unary_with_individuals`` does test this for correctness @petrelharp, doesn't it? I don't think we need to go down the implement-in-Python path, it's quite a simple addition, and the tests do fail if it's not implemented correctly (right @hyanwong?)

**Comment by hyanwong** [comment]
> the tests do fail if it's not implemented correctly (right @hyanwong?)

Yes, they do fail if its not implemented, or not implemented correctly.

**Comment by jeromekelleher** [comment]
As a follow up,  we should open an issue to track adding the functionality in Python, though, so we don't forget about it.

**Comment by hyanwong** [comment]
> As a follow up, we should open an issue to track adding the functionality in Python, though, so we don't forget about it.

Now at https://github.com/tskit-dev/tskit/issues/1120

**Comment by petrelharp** [comment]
> But I have tested a (admittedly single tree) case where one unary node has an individual and another one doesn't, and I've checked the number of nodes that are expected to remain after the KEEP_UNARY_IN_INDIVIDUALS flag. What other tests might you be thinking of?

Well, asyou say, there's tests for (a) equality of tables where the behavior is the same as KEEP_UNARY (b) equality of tables where it's the same as just plain simplify, and (c) for [equality of number of rows](https://github.com/tskit-dev/tskit/pull/1119/files#diff-ba4ed4f290b38255241a3267411ce136dd44b6477cbe0cbb3c39341948fd96f6R1054) where KEEP_UNARY_IN_INDIVIDUALS has new behavior. At minimum I'd compare to the tables that are expected, rather than just counting the number of rows.  It'd also be nice to put in a few more odd situations, like an individual on a unary node above the root; an individual on a unary node above a single sample, unary nodes without individuals in both situations, a case where a node was not unary but becomes unary because of simplification, and a case where an individual has two nodes, and one will be retained because of this but the other not. Like so:

ex

```
    const char *nodes = "1  0   0   -1\n"
                        "1  0   0   0\n"
                        "0  1   0   -1\n"
                        "0  1   0   1\n"
                        "0  2   0   -1\n"
                        "0  3   0   -1\n"
                        "0  3   0   2\n"
                        "0  1   0   -1\n"
                        "0  1   0   3\n"
                        "0  0   0   0\n"
                        "0  0   0   0\n"
                        "0  1   0   3\n";
    const char *edges = "0  2   2   0\n"
                        "0  2   3   1\n"
                        "0  2   4   2,3\n"
                        "0  1   5   4\n"
                        "0  1   6   4\n"
                        "2  3   7   0\n"
                        "2  3   8   1,9\n"
                        "2  3   11   10\n";
    const char *individuals = "0    2.5\n"
                              "0    1.5,3.1\n"
                              "0    1.1\n"
                              "0    1.9\n";
```

**Comment by hyanwong** [comment]
> > But I have tested a (admittedly single tree) case where one unary node has an individual and another one doesn't, and I've checked the number of nodes that are expected to remain after the KEEP_UNARY_IN_INDIVIDUALS flag. What other tests might you be thinking of?
> At minimum I'd compare to the tables that are expected, rather than just counting the number of rows. 

Yes, fair point. Or maybe, simply test that the nodes are the expected ones (e.g. return the new node ids)

> It'd also be nice to put in a few more odd situations, like an individual on a unary node above the root; an individual on a unary node above a single sample, unary nodes without individuals in both situations, 

Shouldn't we then test the normal KEEP_UNARY in this way too. Or if we do (I haven't checked), why would it be any different with KEEP_UNARY_IN_INDIVIDUALS

> a case where a node was not unary but becomes unary because of simplification, and a case where an individual has two nodes, and one will be retained because of this but the other not. 

Yep, I can see that I guess.

**Comment by petrelharp** [comment]
I think this is the expected result:
```
    const char *nodes = "1  0   0   -1\n"
                        "1  0   0   0\n"
                        "0  1   0   1\n"
                        "0  1   0   2\n"
                        "0  2   0   -1\n"
                        "0  3   0   3\n"
    const char *edges =
                        "0  2   2   1\n"
                        "2  3   3   1\n"
                        "0  2   4   0,2\n"
                        "1  2   5   4\n";
    const char *individuals = "0    2.5\n"
                              "0    1.5,3.1\n"
                              "0    1.9\n"
                              "0    1.1\n";
```

**Comment by petrelharp** [comment]
That wasn't quite right, but I've got it sorted out now.

**Comment by jeromekelleher** [comment]
Is this ready to go @petrelharp? It would be good to get this in and ship the next C API release so we can get this into SLiM and so we don't need to hold up #1125.

(Although, tracking individual parents properly would be very nice to be able to do in SLiM too - would it be worth pulling all this into a C API release and updating SLiM then?)

**Comment by hyanwong** [comment]
I think we concluded in https://github.com/MesserLab/SLiM/issues/139#issuecomment-762900817 that on the principle of least harm we would default to setting KEEP_UNARY_IN_INDIVIDUALS to `false` by default in SLiM simplification. So SLiM 3.5.1 is not actually waiting up on this. Perhaps that argues for waiting until the individual parents stuff is merged too?

**Comment by petrelharp** [comment]
I think this is ready to merge - I was just waiting for someone (@hyanwong?) to look at my additional test case. I think it's good, though - merge if you like. We should discuss the C/SLiM release on slack.

**Comment by hyanwong** [comment]
>  I was just waiting for someone (@hyanwong?) to look at my additional test case. 

Oh sorry. I didn't realise you were waiting on me. Merge away.

**Review by benjeffery — APPROVED** [review]

**Comment by benjeffery** [comment]
I am going to squash the commits and merge.

**Comment by bhaller** [comment]
> I think we concluded in [MesserLab/SLiM#139 (comment)](https://github.com/MesserLab/SLiM/issues/139#issuecomment-762900817) that on the principle of least harm we would default to setting KEEP_UNARY_IN_INDIVIDUALS to `false` by default in SLiM simplification. So SLiM 3.5.1 is not actually waiting up on this. Perhaps that argues for waiting until the individual parents stuff is merged too?

As I understood the situation, SLiM 3.5.1 is waiting on this because it seemed better not to have to roll a 3.5.2 release with this feature, or to have the feature delayed until SLiM 3.6 (whatever that turns out to be) happens, which will probably be months down the road.  Anyway, I'm not sure where the right place is now to discuss this with you and @petrelharp – probably not on a merged PR! – but let's figure this out.  :->

**Comment by hyanwong** [comment]
> As I understood the situation, SLiM 3.5.1 is waiting on this 

Actually, I think it's not. It's only waiting on my reworking of the pyslim docs to include "retaining" individuals. That doesn't need this PR in tskit.

> Anyway, I'm not sure where the right place is now to discuss this with you and @petrelharp – probably not on a merged PR! 

Indeed. I think the link above, https://github.com/MesserLab/SLiM/issues/139 is fine.

**Comment by benjeffery** [comment]
Is that a definitive answer? Happy to tag a release if SLiM does need this. Next release after that will likely be a couple of weeks looking at the milestone.

**Comment by hyanwong** [comment]
No hurry for a release from me any more (now that Peter and Ben argued me out of setting `KEEP_UNARY_IN_INDIVIDUALS` as the default). So punt it off for a couple of weeks and go and get yourself a G&T instead.

**Comment by bhaller** [comment]
@hyanwong I don't understand your read on the situation.  Could you explain why what I wrote above is incorrect?

**Comment by bhaller** [comment]
Ah, I see the discussion has moved to Slack...

**Pull request merged**
